# Ejercicios ensembling
En este ejercicio vas a realizar prediciones sobre un dataset de ciudadanos indios diabéticos. Se trata de un problema de clasificación en el que intentaremos predecir 1 (diabético) 0 (no diabético).

### 1. Carga las librerias que consideres comunes al notebook

In [83]:
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold, cross_val_score
from sklearn.model_selection import train_test_split, GridSearchCV

### 2. Lee los datos de [esta direccion](https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv)
Los nombres de columnas son:
```Python
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
```

In [84]:
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
diabetes = pd.read_csv(url, names=names)

In [85]:
diabetes.head()

,preg,plas,pres,skin,test,mass,pedi,age,class
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


In [86]:
diabetes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   preg    768 non-null    int64  
 1   plas    768 non-null    int64  
 2   pres    768 non-null    int64  
 3   skin    768 non-null    int64  
 4   test    768 non-null    int64  
 5   mass    768 non-null    float64
 6   pedi    768 non-null    float64
 7   age     768 non-null    int64  
 8   class   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


### 3. Bagging
Para este apartado tendrás que crear un ensemble utilizando la técnica de bagging ([BaggingClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html)), mediante la cual combinarás 100 [DecisionTreeClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html). Recuerda utilizar también [cross validation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html) con 10 kfolds.

**Para este apartado y siguientes, no hace falta que dividas en train/test**, por hacerlo más sencillo. Simplemente divide tus datos en features y target.

Establece una semilla

In [87]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier

In [88]:
X = diabetes.drop(columns=['class'])
y = diabetes['class']

In [89]:
kf = KFold(
    n_splits=10,
    shuffle=True,
    random_state=42
)

In [90]:
estimator = DecisionTreeClassifier(random_state=42)

bag_clf = BaggingClassifier(
    estimator=estimator,
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

scores = cross_val_score(bag_clf, X, y, cv=kf)
bag_score = np.mean(scores)

print("Scores por fold:", scores)
print("Media accuracy:", np.mean(scores))
print("Std accuracy:", np.std(scores))

Scores por fold: [0.72727273 0.75324675 0.7012987  0.85714286 0.79220779 0.71428571
 0.77922078 0.76623377 0.67105263 0.80263158]
Media accuracy: 0.7564593301435407
Std accuracy: 0.05209454512691592


### 4. Random Forest
En este caso entrena un [RandomForestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) con 100 árboles y un `max_features` de 3. También con validación cruzada

In [91]:
from sklearn.ensemble import RandomForestClassifier

In [92]:
rnd_clf = RandomForestClassifier(
    n_estimators=100,
    max_features=3,
    random_state=42,
    n_jobs=-1
)

scores = cross_val_score(rnd_clf, X, y, cv=kf)
rnd_score = np.mean(scores)

print("Scores por fold:", scores)
print("Media accuracy:", np.mean(scores))
print("Std accuracy:", np.std(scores))

Scores por fold: [0.72727273 0.77922078 0.74025974 0.85714286 0.84415584 0.72727273
 0.79220779 0.74025974 0.67105263 0.81578947]
Media accuracy: 0.7694634313055365
Std accuracy: 0.05565587112247596


### 5. AdaBoost
Implementa un [AdaBoostClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.AdaBoostClassifier.html) con 30 árboles.

In [93]:
from sklearn.ensemble import AdaBoostClassifier

In [94]:
ada_clf = AdaBoostClassifier(
    n_estimators=30,
    random_state=42
)

scores = cross_val_score(ada_clf, X, y, cv=kf)
ada_score = np.mean(scores)

print("Scores por fold:", scores)
print("Media accuracy:", np.mean(scores))
print("Std accuracy:", np.std(scores))

Scores por fold: [0.74025974 0.79220779 0.72727273 0.77922078 0.83116883 0.67532468
 0.79220779 0.77922078 0.68421053 0.81578947]
Media accuracy: 0.7616883116883117
Std accuracy: 0.050336101925772855


### 6. GradientBoosting
Implementa un [GradientBoostingClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html) con 100 estimadores

In [95]:
from sklearn.ensemble import GradientBoostingClassifier

In [96]:
gbct = GradientBoostingClassifier(
    n_estimators=100,
    random_state=42
)

scores = cross_val_score(gbct, X, y, cv=kf)
gbct_score = np.mean(scores)

print("Scores por fold:", scores)
print("Media accuracy:", np.mean(scores))
print("Std accuracy:", np.std(scores))

Scores por fold: [0.72727273 0.79220779 0.74025974 0.85714286 0.81818182 0.72727273
 0.81818182 0.74025974 0.65789474 0.77631579]
Media accuracy: 0.765498974709501
Std accuracy: 0.05527669626375457


### 7. XGBoost
Para este apartado utiliza un [XGBoostClassifier](https://docs.getml.com/latest/api/getml.predictors.XGBoostClassifier.html) con 100 estimadores. XGBoost no forma parte de la suite de modelos de sklearn, por lo que tendrás que instalarlo con pip install

In [97]:
from xgboost import XGBClassifier

In [98]:
xgb_clas = XGBClassifier(
    n_estimators=100,
    random_state=42,
    eval_metric="logloss",   # evita warnings
    n_jobs=-1
)

scores = cross_val_score(xgb_clas, X, y, cv=kf)
xgb_score = np.mean(scores)

print("Scores por fold:", scores)
print("Media accuracy:", np.mean(scores))
print("Std accuracy:", np.std(scores))

Scores por fold: [0.68831169 0.67532468 0.71428571 0.81818182 0.80519481 0.71428571
 0.76623377 0.76623377 0.68421053 0.77631579]
Media accuracy: 0.7408578263841422
Std accuracy: 0.049335739322384435


### 8. Primeros resultados
Crea un dataframe con los resultados y sus algoritmos, ordenándolos de mayor a menor

In [99]:
resultados = pd.DataFrame({
    "modelo": [
        "Bagging (100 árboles)",
        "Random Forest (100 árboles, max_features=3)",
        "AdaBoost (30 árboles)",
        "GradientBoosting (100 estimadores)",
        "XGBoost (100 estimadores)"
    ],
    "accuracy": [
        bag_score,
        rnd_score,
        ada_score,
        gbct_score,
        xgb_score
    ]
})

# Ordenar de mayor a menor accuracy
resultados = resultados.sort_values(by="accuracy", ascending=False).reset_index(drop=True)

In [100]:
resultados

,modelo,accuracy
0,"Random Forest (100 árboles, max_features=3)",0.769463
1,GradientBoosting (100 estimadores),0.765499
2,AdaBoost (30 árboles),0.761688
3,Bagging (100 árboles),0.756459
4,XGBoost (100 estimadores),0.740858


### 9. Hiperparametrización
Vuelve a entrenar los modelos de nuevo, pero esta vez dividiendo el conjunto de datos en train/test y utilizando un gridsearch para encontrar los mejores hiperparámetros.

In [101]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y  # si tu problema es clasificación, mejor así
)

In [102]:
bag_param = {
    "n_estimators": [50, 100, 200],
    "max_samples": [0.7, 1.0],
    "random_state": [42]
}

In [103]:
bag_gs = GridSearchCV(BaggingClassifier(
    estimator=DecisionTreeClassifier(random_state=42)),
    param_grid=bag_param,
    cv=5,
    scoring="accuracy",
    n_jobs=-1
)

bag_gs.fit(X_train, y_train)

best_bag = bag_gs.best_estimator_
bag_test_score = best_bag.score(X_test, y_test)

print("Mejores hiperparámetros:", bag_gs.best_params_)
print("Mejor score (CV):", bag_gs.best_score_)
print("Accuracy en test:", bag_test_score)

Mejores hiperparámetros: {'max_samples': 0.7, 'n_estimators': 50, 'random_state': 42}
Mejor score (CV): 0.7703985072637611
Accuracy en test: 0.7662337662337663


In [104]:
rnd_param = {
    "n_estimators": [100, 200],
    "max_features": [3, "sqrt", "log2"],
    "max_depth": [None, 5, 10],
    "random_state": [42]
}

In [105]:
rnd_gs = GridSearchCV(RandomForestClassifier(random_state=42),
    param_grid=rnd_param,
    cv=5,
    scoring="accuracy",
    n_jobs=-1
)

rnd_gs.fit(X_train, y_train)

best_rnd = rnd_gs.best_estimator_
rnd_test_score = best_rnd.score(X_test, y_test)

print("Mejores hiperparámetros:", rnd_gs.best_params_)
print("Mejor score (CV):", rnd_gs.best_score_)
print("Accuracy en test:", rnd_test_score)

Mejores hiperparámetros: {'max_depth': None, 'max_features': 3, 'n_estimators': 100, 'random_state': 42}
Mejor score (CV): 0.776915900306544
Accuracy en test: 0.7597402597402597


In [106]:
ada_param = {
    "n_estimators": [30, 50, 100],
    "learning_rate": [0.01, 0.1, 1.0],
    "random_state": [42]
}

In [107]:
ada_gs = GridSearchCV(AdaBoostClassifier(random_state=42),
    param_grid=ada_param,
    cv=5,
    scoring="accuracy",
    n_jobs=-1
)

ada_gs.fit(X_train, y_train)

best_ada = ada_gs.best_estimator_
ada_test_score = best_ada.score(X_test, y_test)

print("Mejores hiperparámetros:", ada_gs.best_params_)
print("Mejor score (CV):", ada_gs.best_score_)
print("Accuracy en test:", ada_test_score)

Mejores hiperparámetros: {'learning_rate': 1.0, 'n_estimators': 30, 'random_state': 42}
Mejor score (CV): 0.767093162734906
Accuracy en test: 0.7857142857142857


In [108]:
gbct_param = {
    "n_estimators": [100, 200],
    "learning_rate": [0.05, 0.1],
    "max_depth": [3, 5],
    "random_state": [42]
}

In [109]:
gbct_gs = GridSearchCV(GradientBoostingClassifier(random_state=42),
    param_grid=gbct_param,
    cv=5,
    scoring="accuracy",
    n_jobs=-1
)

gbct_gs.fit(X_train, y_train)

best_gbct = gbct_gs.best_estimator_
gbct_test_score = best_gbct.score(X_test, y_test)

print("Mejores hiperparámetros:", gbct_gs.best_params_)
print("Mejor score (CV):", gbct_gs.best_score_)
print("Accuracy en test:", gbct_test_score)

Mejores hiperparámetros: {'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100, 'random_state': 42}
Mejor score (CV): 0.7541516726642676
Accuracy en test: 0.7532467532467533


In [110]:
xgb_param = {
    "n_estimators": [100, 200],
    "max_depth": [3, 5],
    "learning_rate": [0.05, 0.1],
    "random_state": [42]
}

In [111]:
xgb_gs = GridSearchCV(XGBClassifier(
            random_state=42,
            eval_metric="logloss",
            n_jobs=-1
        ),
    param_grid=xgb_param,
    cv=5,
    scoring="accuracy",
    n_jobs=-1
)

xgb_gs.fit(X_train, y_train)

best_xgb = xgb_gs.best_estimator_
xgb_test_score = best_xgb.score(X_test, y_test)

print("Mejores hiperparámetros:", xgb_gs.best_params_)
print("Mejor score (CV):", xgb_gs.best_score_)
print("Accuracy en test:", xgb_test_score)

Mejores hiperparámetros: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100, 'random_state': 42}
Mejor score (CV): 0.7574037051845928
Accuracy en test: 0.7532467532467533


### 10. Conclusiones finales

In [112]:
resultados_2 = pd.DataFrame({
    "modelo": [
        "Bagging",
        "Random",
        "AdaBoost",
        "GradientBoosting",
        "XGBoost"
    ],
    "accuracy_test": [
        bag_test_score,
        rnd_test_score,
        ada_test_score,
        gbct_test_score,
        xgb_test_score
    ]
})

resultados_2 = resultados_2.sort_values(by="accuracy_test", ascending=False).reset_index(drop=True)

In [115]:
resultados

,modelo,accuracy
0,"Random Forest (100 árboles, max_features=3)",0.769463
1,GradientBoosting (100 estimadores),0.765499
2,AdaBoost (30 árboles),0.761688
3,Bagging (100 árboles),0.756459
4,XGBoost (100 estimadores),0.740858


In [116]:
resultados_2

,modelo,accuracy_test
0,AdaBoost,0.785714
1,Bagging,0.766234
2,Random,0.759740
3,GradientBoosting,0.753247
4,XGBoost,0.753247


Tras aplicar GridSearchCV y evaluar los modelos el modelo con mejor rendimiento fue AdaBoost, alcanzando un accuracy de 0.7857. El uso de train/test y la búsqueda de hiperparámetros muestra que AdaBoost generaliza mejor en este dataset.